analyzers/src/SonarAnalyzer.Common/Helpers/KnownType.cs

```diff
@@ -19,589 +19,136 @@
+            internal RegularKnownType(string nameSpace, string typeName, bool isArray, params string[] genericArguments)
+            {
+                this.nameSpace = nameSpace;
+                this.typeName = typeName;
+                this.isArray = isArray;
+                this.genericArguments = genericArguments ?? Array.Empty<string>();
+            }
 
-#pragma warning restore S103    // Lines should not be too long
-#pragma warning restore SA1310  // FieldNamesMustNotContainUnderscore
-#pragma warning restore SA1311  // Static readonly fields should begin with upper-case letter
-#pragma warning restore SA1307  // Field 'log4net_Config_XmlConfigurator' should begin with upper-case letter
-#pragma warning restore SA1304  // Non-private readonly fields should begin with upper-case letter
+            // ToDo: Temporary fix to avoid changing all the 342 usages in this PR. Should be removed later.
+            internal RegularKnownType(string fullTypeName)
+            {
+                var lastIndex = fullTypeName.LastIndexOf('.');
+                if (lastIndex > 0)
+                {
+                    nameSpace = fullTypeName.Substring(0, lastIndex);
+                    typeName = fullTypeName.Substring(lastIndex + 1);
+                }
+                else
+                {
+                    typeName = fullTypeName;
+                    nameSpace = string.Empty;
+                }
 
-        private readonly bool isSpecialType;
-        private readonly SpecialType specialType;
-        private readonly Lazy<string> shortName;
+                genericArguments = Array.Empty<string>();
+            }
 
-        public string TypeName { get; }
-        public string ShortName => shortName.Value;
+            internal override bool Matches(ITypeSymbol symbol) =>
+                IsMatch(symbol)
+                || IsMatch(symbol.OriginalDefinition);
 
-        private KnownType(string typeName) : this(SpecialType.None, typeName) { }
+            private bool IsMatch(ITypeSymbol symbol)
+            {
+                switch (symbol)
+                {
+                    case null:
+                        throw new ArgumentNullException(nameof(symbol));
 
-        private KnownType(SpecialType specialType, string typeName)
-        {
-            this.specialType = specialType;
-            TypeName = typeName;
-            shortName = new Lazy<string>(() => typeName.Split('.').Last());
-            isSpecialType = specialType != SpecialType.None;
+                    case IArrayTypeSymbol arrayTypeSymbol:
+                        return isArray
+                               && arrayTypeSymbol.ElementType.Name == typeName;
+
+                    case INamedTypeSymbol { IsGenericType: true } genericTypeSymbol when genericTypeSymbol.TypeArguments.Length != genericArguments.Length:
+                        return false;
+
+                    case INamedTypeSymbol { IsGenericType: true } genericTypeSymbol:
```

Comment:
Then we can remove the previous check for `Length` too I think

Reply:
That's a good point. Let's test the class name first to cheaply kill most of the candidates. And then use Linq for narrow candidates.

Because most of the arguments are `T` and will match.

Reply:
We also don't support generic array now (doesn't make sense, not really ended, but it's a fact not visible from the contract).
If we receive something else then `INamedTypeSymbol` and require `genericParameters`, then we would return `true` even when the result should be false.

What about this to address it?
```
            private bool IsMatch(ITypeSymbol symbol)
            {
                _ = symbol ?? throw new ArgumentNullException(nameof(symbol));
                if (IsArray)
                {
                    if (symbol is IArrayTypeSymbol array)
                    {
                        symbol = array.ElementType;
                    }
                    else
                    {
                        return false;
                    }
                }
                return symbol.Name == TypeName
                    && NamespaceMatches(symbol)
                    && (!genericParameters.Any() || (symbol is INamedTypeSymbol namedType && namedType.TypeArguments.Select(x => x.Name).SequenceEqual(genericParameters)));
            }
```
I'm not very happy with how the last condition is complicated, but it can start a discussion.

The idea is to offload the array element type for further processing. To force generic arguments when required. And compare them by linq for simplicity.

Reply:
This gets too complicated for a scenario we don't need. I suggest to let this for when we actually need it.

Reply:
Thanks!